src/controller/proxy/controller.go

```diff
@@ -0,0 +1,230 @@
+	}
+
+	if err != nil {
+		if errors.IsNotFoundErr(err) && len(art.Tag) > 0 {
+			go func() {
+				c.local.DeleteManifest(ctx, repo, string(art.Tag))
+			}()
+		}
+		serror.SendError(w, err)
+		return err
+	}
+
+	ct, payload, err := man.Payload()
+	setHeaders(w, int64(len(payload)), ct, art.Digest)
+	w.Write(payload)
+
+	// Push manifest in background
+	go func() {
+		c.waitAndPushManifest(ctx, p, repo, string(art.Tag), man, art, ct, r)
+	}()
+
+	return nil
+}
+
+func (c *controller) ProxyBlob(ctx context.Context, p *models.Project, repo string, dig string, w http.ResponseWriter, r RemoteInterface) error {
+	log.Debugf("The blob doesn't exist, proxy the request to the target server, url:%v", repo)
+	desc := distribution.Descriptor{}
+	size, bReader, err := r.BlobReader(repo, dig)
+	defer bReader.Close()
+	if err != nil {
+		log.Errorf("failed to pull blob, error %v", err)
+		return err
+	}
+	written, err := io.CopyN(w, bReader, size)
+	if written != size {
+		e := errors.Errorf("The size mismatch, actual:%d, expected: %d", written, size)
+		return e
+	}
+	desc.Size = size
+	desc.Digest = digest.Digest(dig)
+
+	setHeaders(w, size, desc.MediaType, dig)
+	go func() {
+		err := c.putBlobToLocal(ctx, p, repo, p.Name+"/"+repo, desc, r)
+		if err != nil {
+			log.Errorf("error while putting blob to local, %v", err)
+		}
+	}()
+	return nil
+}
+
+// NewController create an instance of the Controller
+func NewController() Controller {
+	return &controller{
+		blobCtl:     blob.Ctl,
+		registryMgr: registry.NewDefaultManager(),
+		inflight:    make(map[string]interface{}),
+		artifactCtl: artifact.Ctl,
+		local:       CreateLocalInterface(),
+	}
+}
+
+func (c *controller) putBlobToLocal(ctx context.Context, p *models.Project, orgRepo string, localRepo string, desc distribution.Descriptor, r RemoteInterface) error {
+	log.Debugf("Put blob to local registry!, sourceRepo:%v, localRepo:%v, digest: %v", orgRepo, localRepo, desc.Digest)
+	_, bReader, err := r.BlobReader(orgRepo, string(desc.Digest))
+	defer bReader.Close()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	err = c.local.PushBlob(ctx, p, localRepo, desc, bReader)
+	return err
+}
+
+func setHeaders(w http.ResponseWriter, size int64, mediaType string, dig string) {
+	w.Header().Set("Content-Length", fmt.Sprintf("%v", size))
+	if len(mediaType) > 0 {
+		w.Header().Set("Content-Type", mediaType)
+	}
+	w.Header().Set("Docker-Content-Digest", dig)
+	w.Header().Set("Etag", dig)
+}
+
+func (c *controller) waitAndPushManifest(ctx context.Context, p *models.Project, repo, tag string, man distribution.Manifest, art lib.ArtifactInfo, contType string, r RemoteInterface) {
+	localRepo := art.ProjectName + "/" + repo
+	if contType == manifestlist.MediaTypeManifestList {
+		err := c.local.PushManifestList(ctx, p, localRepo, tag, art, man)
+		if err != nil {
+			log.Errorf("error when push manifest list to local:%v", err)
+		}
+		return
+	}
+
+	for n := 0; n < maxWait; n = n + 1 {
```

Comment:
So we just wait for 200 seconds in total, is it correct?

Reply:
Will updated, keep it for debug quickly

Reply:
Will updated it before release, keep it for debug quickly